library/L1_Peripheral/msp432p401r/system_controller.hpp

```diff
@@ -5,44 +5,421 @@
+                                     select_value,
+                                     kPrimaryClockSelectMasks[Value(clock)]);
+  }
+  /// Determines the target frequency range select value and the DCO tuning
+  /// value necessary to generate a desired target frequency.
+  ///
+  /// @see 6.2.8.3 DCO Ranges and Tuning in the MSP432P4xx Reference Manual.
+  ///      https://www.ti.com/lit/ug/slau356i/slau356i.pdf#page=386
+  ///
+  /// @param frequency The target DCO frequency.
+  /// @returns A structure containing the DCO configuration values.
+  DcoConfiguration_t CalculateDcoConfiguration(
+      units::frequency::kilohertz_t frequency) const
   {
+    constexpr units::frequency::megahertz_t kMinimumDcoSpeed = 1_MHz;
+    constexpr units::frequency::megahertz_t kMaximumDcoSpeed = 48_MHz;
+    SJ2_ASSERT_FATAL(
+        (kMinimumDcoSpeed <= frequency) && (frequency <= kMaximumDcoSpeed),
+        "The desired frequency must be between 1 MHz and 48 MHz.");
+    // Determine the DCO tuning configuration values by finding the DCO
+    // frequency range, DCO constant, and DCO calibration values based on the
+    // desired target frequency.
+    constexpr units::frequency::kilohertz_t kDcoCenterFrequencies[] = {
+      1500_kHz, 3_MHz, 6_MHz, 12_MHz, 24_MHz, 48_MHz
+    };
+    uint8_t dco_frequency_select = 0b000;
+    float dco_constant =
+        static_cast<float>(device_descriptors->DCOIR_CONSTK_RSEL04);
+    int32_t dco_calibration =
+        static_cast<int32_t>(device_descriptors->DCOIR_FCAL_RSEL04);
+
+    if (frequency >= 32_MHz)
+    {
+      // target frequency is 32 MHz to 48 MHz
+      dco_frequency_select = 0b101;
+      dco_constant = static_cast<float>(device_descriptors->DCOIR_CONSTK_RSEL5);
+      dco_calibration =
+          static_cast<int32_t>(device_descriptors->DCOIR_FCAL_RSEL5);
+    }
+    else if (frequency >= 16_MHz)
+    {
+      // target frequency is 16 MHz to 32 MHz
+      dco_frequency_select = 0b100;
+    }
+    else if (frequency >= 8_MHz)
+    {
+      // target frequency is 8 MHz to 16 MHz
+      dco_frequency_select = 0b011;
+    }
+    else if (frequency >= 4_MHz)
+    {
+      // target frequency is 4 MHz to 8 MHz
+      dco_frequency_select = 0b010;
+    }
+    else if (frequency >= 2_MHz)
+    {
+      // target frequency is 2 MHz to 4 MHz
+      dco_frequency_select = 0b001;
+    }
+    else
+    {
+      // target frequency is 1 MHz to 2 MHz with center frequency of 1.5 MHz
+      // use the default initialized values
+    }
+    // Calculate the signed 10-bit tuning value using Equation 6 from
+    // https://www.ti.com/lit/ug/slau356i/slau356i.pdf#page=387
+    const auto kCenterFrequency = kDcoCenterFrequencies[dco_frequency_select];
+    const int16_t kTuneValue    = static_cast<int16_t>(
+        (((frequency - kCenterFrequency).to<float>() *
+          (1.0f + dco_constant * static_cast<float>(768 - dco_calibration))) /
+         (frequency.to<float>() * dco_constant)));
```

Comment:
Can you split this up into multiple parts. Its really difficult to read whats happening.